start.sh: Try to load the in-kernel exfat driver if needed

jprvita opened the pull request:
Prior to EOS 4 exFAT support was provided via FUSE. Since the initrd
does not contain the FUSE driver or the kernel module, it will not be
possible to mount the USB device containing the image once we pivot to
the initrd.

One way around this problem is to manually load the in-kernel exfat
driver before pivot-root, so it is available in the running kernel after
pivot-root (which is also how this works for any other filesystem we
support in the kernel).

https://phabricator.endlessm.com/T31536

jprvita committed:
start.sh: Drop unused variable KERNEL_VERSION

This was a leftover from commit "591ef01 start.sh: hardcodes initramfs
path".

jprvita committed:
start.sh: Try to load the in-kernel exfat driver if needed

Prior to EOS 4.0 exFAT support was provided via FUSE. Since the initrd
does not contain the FUSE driver or the kernel module, it will not be
possible to mount the USB device containing the image once we pivot to
the initrd.

One way around this problem is to manually load the in-kernel exfat
driver before pivot-root, so it is available in the running kernel after
pivot-root (which is also how this works for any other filesystem we
support in the kernel).

This commit tried to load the in-kernel exfat driver if it detects any
exFAT filesystem attached to the system. This is the best we can do from
start.sh, as at the point it is running it has already been copied from
the USB device to /var by eos-factory-test-helper.service.

https://phabricator.endlessm.com/T31536

wjt commented on eos-usb-reflash/factory-test/start.sh:
```diff
@@ -6,7 +6,29 @@
+    if [ $exfat_loaded -ne 0 ] ; then
+        echo -n "This version of EOS does not have in-kernel exFAT support. "
+        echo -n "You need a device formatted with a compatible filesystem "
+        echo "(ext2 / ext4) to reflash this computer."
```
Is regular old fat32 supported? That's an in-kernel module that's been around forever.

jprvita replied:
FAT32 should technically work (I haven't tested it), but it has a file size limitation of 4GB, so in practice we won't be able to use it with any image except for the `base` personality (already 2.6GB) atm. Assuming it works, do you think we should also mention it on the comment?
```suggestion
        echo "(ext2 / ext4, or FAT32 if your image file has less than 4GB) to reflash this computer."
```

wjt replied:
Great point… probably not worth mentioning then. I would expect that users of this tool are more likely to be using more substantial images.

jprvita commented:
Updated to a more reasonable timeout (30s), removed the test suite from /var so it does not start again on the following boot, and triggered a shutdown instead of exiting to limbo.

wjt commented on eos-usb-reflash/factory-test/start.sh:
```diff
@@ -6,6 +6,30 @@
+# the exfat kernel module, as on EOS prior to 4.0 the USB device has been
+# mounted using the FUSE exFAT driver, and the initramfs we are going to pivot
+# to does not contain an exFAT driver (neither in-kernel nor via FUSE).
+FSTYPES=$(lsblk --noheadings --output FSTYPE)
```
Do we care if there is an exFAT filesystem on a device other than `$mydir`? Here you're checking all devices, but I think only the one that holds the initramfs matters. Would it work to find the mountpoint for `$mydir` and run `$lsblk` against that? (I guess in practice there's no difference.)

wjt replied:
Oh… is `$mydir` the directory in `/var` that everything's already been unpacked to, so actually we don't know what USB device will need to be mounted?

jprvita replied:
> Oh… is `$mydir` the directory in `/var` that everything's already been unpacked to, so actually we don't know what USB device will need to be mounted?

Exactly! That is why I am checking if any exfat fs is attached to the system, I think it is a good enough heuristic in this situation.

wjt commented on eos-usb-reflash/factory-test/start.sh:
```diff
@@ -6,6 +6,30 @@
+if [[ "$FSTYPES" =~ "exfat" ]] ; then
+    echo "Detected an exFAT filesystem attached to this machine;"
+    echo "Trying to load the 'exfat' kernel module..."
+    set +e
+    modprobe exfat
+    exfat_loaded=$?
+    if [ $exfat_loaded -ne 0 ] ; then
```
You don't need to `set +e`; you can use:

```
if ! modprobe exfat ; then
...
fi
```

jprvita replied:
Good point!

jprvita committed:
Update USB tarballs after latest changes

wjt reviewed (approved):
LGTM if it works

jprvita commented:
> LGTM if it works

Worked as expected on eos3.7 (took the error path from exfat, reflashed from ext4) and eos3.9 (reflashed from exfat). Merging per the CR+ above.
